#第三回
##講師
- B2 好田

##TA
- B3 勝又
- B3 戸倉





In [ ]:
#環境構築用のセル

#PostgreSQL13 のインストールとサーバーの起動
!sudo cat > /etc/apt/sources.list.d/pgdg.list <<< "deb http://apt.postgresql.org/pub/repos/apt/ $(lsb_release -cs)-pgdg main"
!sudo curl https://www.postgresql.org/media/keys/ACCC4CF8.asc | sudo apt-key add -  > /dev/null
!sudo apt-get -y -qq update > /dev/null
!sudo apt-get -y -qq install postgresql-13 > /dev/null
!sudo pg_ctlcluster 13 main start > /dev/null
!sudo -u postgres psql -U postgres --version

#Pagilaデータベースの作成、ユーザーpadwanの設定
!sudo -u postgres psql -U postgres -c "create database pagila;"
!sudo -u postgres psql -U postgres -c "create user padwan with superuser;"
!sudo -u postgres psql -U postgres -c "alter user padwan password 'padwan12345';"
!sudo -u postgres psql -U postgres -c "grant all privileges on database pagila to padwan;"
!sudo -u postgres psql -U postgres -c "grant pg_read_server_files to padwan;"
!sudo -u postgres psql -U postgres -c "grant pg_write_server_files to padwan;"

#PagilaのGitHubリポジトリをcloneし、Pagila設定用SQLファイルを実行する
!sudo mkdir /tmp/data
!git clone https://github.com/devrimgunduz/pagila.git /tmp/data/pagila
base = "/tmp/data/pagila/"
sql_names = ('pagila-schema.sql', 'pagila-data.sql')
for sql_name in sql_names:
   SQLFILE=base+sql_name
   !sudo cat $SQLFILE | sudo -u postgres psql -U postgres -d pagila
#SQL入力準備（以降は%%sqlに続く行にSQLを入力可能）
%load_ext sql
%config SqlMagic.autopandas = True
%sql postgresql://padwan:padwan12345@localhost:5432/pagila
!git clone https://github.com/appacademy/aa-sql-sql-zoo.git
!sudo -u postgres psql -U postgres -d pagila -a -f /content/aa-sql-sql-zoo/data/create_tables.sql

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Warning: apt-key is deprecated. Manage keyring files in trusted.gpg.d instead (see apt-key(8)).
100  4812  100  4812    0     0   7571      0 --:--:-- --:--:-- --:--:--  7577
W: http://apt.postgresql.org/pub/repos/apt/dists/jammy-pgdg/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 12.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend:

#この環境で使えるテーブル
- film_actor
- address
- city
- actor
- actor_info
- category
- country
- customer
- customer_list
- film_list
- nicer_but_slower_film_list
- film
- film_category
- inventory
- language
- sales_by_film_category
- rental
- sales_by_store
- staff_list
- payment
- store

#復習

問1. film テーブルから、映画の総置換費用（replacement_cost）を評価別に合計し、各評価カテゴリに対してその総合計を取得します

In [ ]:
%%sql


 * postgresql://padwan:***@localhost:5432/pagila
5 rows affected.


sum rating
0  3678.06     PG
1  3945.05      R
2  3582.22      G
3  4549.77  PG-13
4  4228.90  NC-17

問2. film テーブルから、映画のデータを取得し、それらのデータを replacement_cost 列の値に基づいて昇順に並べ替えてください。

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
1000 rows affected.


film_id           title  \
0        888  THIN SAGEBRUSH   
1        656   PAPI NECKLACE   
2        686  PLUTO OLEANDER   
3        886  THEORY MERMAID   
4        260  DUDE BLINDNESS   
..       ...             ...   
995      238    DOCTOR GRAIL   
996      944    VIRGIN DAISY   
997       81   BLINDNESS GUN   
998      994   WYOMING STORM   
999      483   JERICHO MULAN   

                                           description  release_year  \
0    A Emotional Drama of a Husband And a Lumberjac...          2006   
1    A Fanciful Display of a Car And a Monkey who m...          2006   
2    A Action-Packed Reflection of a Car And a Moos...          2006   
3    A Fateful Yarn of a Composer And a Monkey who ...          2006   
4    A Stunning Reflection of a Husband And a Lumbe...          2006   
..                                                 ...           ...   
995  A Insightful Drama of a Womanizer And a Waitre...          2006   
996  A Awe-Inspiring Documentary of a Robot And a M...          2006   
997  A Touching Drama of a Robot And a Dentist who ...          2006   
998  A Awe-Inspiring Panorama of a Robot And a Boat...          2006   
999  A Amazing Yarn of a Hunter And a Butler who mu...          2006   

     language_id original_language_id  rental_duration rental_rate  length  \
0              1                 None                5        4.99      53   
1              1                 None                3        0.99     128   
2              1                 None                5        4.99      84   
3              1                 None                5        0.99     184   
4              1                 None                3        4.99     132   
..           ...                  ...              ...         ...     ...   
995            1                 None                4        2.99      57   
996            1                 None                6        4.99     179   
997            1                 None                6        4.99     103   
998            1                 None                6        4.99     100   
999            1                 None                3        2.99     171   

    replacement_cost rating                      last_update  \
0               9.99  PG-13 2022-09-10 16:46:03.905795+00:00   
1               9.99     PG 2022-09-10 16:46:03.905795+00:00   
2               9.99      R 2022-09-10 16:46:03.905795+00:00   
3               9.99  PG-13 2022-09-10 16:46:03.905795+00:00   
4               9.99      G 2022-09-10 16:46:03.905795+00:00   
..               ...    ...                              ...   
995            29.99      G 2022-09-10 16:46:03.905795+00:00   
996            29.99  PG-13 2022-09-10 16:46:03.905795+00:00   
997            29.99  PG-13 2022-09-10 16:46:03.905795+00:00   
998            29.99  PG-13 2022-09-10 16:46:03.905795+00:00   
999            29.99  NC-17 2022-09-10 16:46:03.905795+00:00   

                                  special_features  \
0                              [Behind the Scenes]   
1    [Trailers, Deleted Scenes, Behind the Scenes]   
2                              [Behind the Scenes]   
3              [Deleted Scenes, Behind the Scenes]   
4                       [Trailers, Deleted Scenes]   
..                                             ...   
995    [Trailers, Commentaries, Behind the Scenes]   
996                  [Trailers, Behind the Scenes]   
997                  [Trailers, Behind the Scenes]   
998                               [Deleted Scenes]   
999                                 [Commentaries]   

                                              fulltext  
0    'ancient':18 'build':14 'cat':16 'drama':5 'em...  
1    'ancient':18 'car':8 'display':5 'escap':14 'f...  
2    'action':5 'action-pack':4 'car':10,18 'moos':...  
3    'compos':8 'fate':4 'first':19 'man':20 'merma...  
4    'abandon':19 'blind':2 'dude':1 'face':14 'fri...  
..                                                 ...  
9

問3. customerテーブルの全行に加え、activeカラムが1の時は‘ACTIVE’, 1でない場合は‘non-acrive’を1カラムとして追加表示してください。

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
599 rows affected.


customer_id  store_id first_name  last_name  \
0              1         1       MARY      SMITH   
1              2         1   PATRICIA    JOHNSON   
2              3         1      LINDA   WILLIAMS   
3              4         2    BARBARA      JONES   
4              5         1  ELIZABETH      BROWN   
..           ...       ...        ...        ...   
594          595         1   TERRENCE  GUNDERSON   
595          596         1    ENRIQUE   FORSYTHE   
596          597         1    FREDDIE     DUGGAN   
597          598         1       WADE   DELVALLE   
598          599         2     AUSTIN    CINTRON   

                                     email  address_id  activebool  \
0            MARY.SMITH@sakilacustomer.org           5        True   
1      PATRICIA.JOHNSON@sakilacustomer.org           6        True   
2        LINDA.WILLIAMS@sakilacustomer.org           7        True   
3         BARBARA.JONES@sakilacustomer.org           8        True   
4       ELIZABETH.BROWN@sakilacustomer.org           9        True   
..                                     ...         ...         ...   
594  TERRENCE.GUNDERSON@sakilacustomer.org         601        True   
595    ENRIQUE.FORSYTHE@sakilacustomer.org         602        True   
596      FREDDIE.DUGGAN@sakilacustomer.org         603        True   
597       WADE.DELVALLE@sakilacustomer.org         604        True   
598      AUSTIN.CINTRON@sakilacustomer.org         605        True   

    create_date               last_update  active category  
0    2022-02-14 2022-02-15 09:57:20+00:00       1   ACTIVE  
1    2022-02-14 2022-02-15 09:57:20+00:00       1   ACTIVE  
2    2022-02-14 2022-02-15 09:57:20+00:00       1   ACTIVE  
3    2022-02-14 2022-02-15 09:57:20+00:00       1   ACTIVE  
4    2022-02-14 2022-02-15 09:57:20+00:00       1   ACTIVE  
..          ...                       ...     ...      ...  
594  2022-02-14 2022-02-15 09:57:20+00:00       1   ACTIVE  
595  2022-02-14 2022-02-15 09:57:20+00:00       1   ACTIVE  
596  2022-02-14 2022-02-15 09:57:20+00:00       1   ACTIVE  
597  2022-02-14 2022-02-15 09:57:20+00:00       1   ACTIVE  
598  2022-02-14 2022-02-15 09:57:20+00:00       1   ACTIVE  

[599 rows x 11 columns]

#1.JOIN

##1.1 INNER JOIN
INNER JOIN(内部結合)はそれぞれのテーブルのしていしたカラム(列)の値が一致するものだけを結合します

```
SELECT
 (*)
FROM
 (Table1)
INNER JOIN
 (Table2)
ON
 (結合の条件)
```

###例：

In [ ]:
%%sql
SELECT
 *
FROM
 film

 * postgresql://padwan:***@localhost:5432/pagila
1000 rows affected.


film_id              title  \
0          1   ACADEMY DINOSAUR   
1          2     ACE GOLDFINGER   
2          3   ADAPTATION HOLES   
3          4   AFFAIR PREJUDICE   
4          5        AFRICAN EGG   
..       ...                ...   
995      996     YOUNG LANGUAGE   
996      997         YOUTH KICK   
997      998       ZHIVAGO CORE   
998      999  ZOOLANDER FICTION   
999     1000          ZORRO ARK   

                                           description  release_year  \
0    A Epic Drama of a Feminist And a Mad Scientist...          2006   
1    A Astounding Epistle of a Database Administrat...          2006   
2    A Astounding Reflection of a Lumberjack And a ...          2006   
3    A Fanciful Documentary of a Frisbee And a Lumb...          2006   
4    A Fast-Paced Documentary of a Pastry Chef And ...          2006   
..                                                 ...           ...   
995  A Unbelieveable Yarn of a Boat And a Database ...          2006   
996  A Touching Drama of a Teacher And a Cat who mu...          2006   
997  A Fateful Yarn of a Composer And a Man who mus...          2006   
998  A Fateful Reflection of a Waitress And a Boat ...          2006   
999  A Intrepid Panorama of a Mad Scientist And a B...          2006   

     language_id original_language_id  rental_duration rental_rate  length  \
0              1                 None                6        0.99      86   
1              1                 None                3        4.99      48   
2              1                 None                7        2.99      50   
3              1                 None                5        2.99     117   
4              1                 None                6        2.99     130   
..           ...                  ...              ...         ...     ...   
995            1                 None                6        0.99     183   
996            1                 None                4        0.99     179   
997            1                 None                6        0.99     105   
998            1                 None                5        2.99     101   
999            1                 None                3        4.99      50   

    replacement_cost rating                      last_update  \
0              20.99     PG 2022-09-10 16:46:03.905795+00:00   
1              12.99      G 2022-09-10 16:46:03.905795+00:00   
2              18.99  NC-17 2022-09-10 16:46:03.905795+00:00   
3              26.99      G 2022-09-10 16:46:03.905795+00:00   
4              22.99      G 2022-09-10 16:46:03.905795+00:00   
..               ...    ...                              ...   
995             9.99      G 2022-09-10 16:46:03.905795+00:00   
996            14.99  NC-17 2022-09-10 16:46:03.905795+00:00   
997            10.99  NC-17 2022-09-10 16:46:03.905795+00:00   
998            28.99      R 2022-09-10 16:46:03.905795+00:00   
999            18.99  NC-17 2022-09-10 16:46:03.905795+00:00   

                                special_features  \
0            [Deleted Scenes, Behind the Scenes]   
1                     [Trailers, Deleted Scenes]   
2                     [Trailers, Deleted Scenes]   
3              [Commentaries, Behind the Scenes]   
4                               [Deleted Scenes]   
..                                           ...   
995                [Trailers, Behind the Scenes]   
996                [Trailers, Behind the Scenes]   
997                             [Deleted Scenes]   
998                   [Trailers, Deleted Scenes]   
999  [Trailers, Commentaries, Behind the Scenes]   

                                              fulltext  
0    'academi':1 'battl':15 'canadian':20 'dinosaur...  
1    'ace':1 'administr':9 'ancient':19 'astound':4...  
2    'adapt':1 'astound':4 'baloon':19 'car':11 'fa...  
3    'affair':1 'chase':14 'documentari':5 'fanci':...  
4    'african':1 'chef':11 'dentist':14 'documentar...  
..                                            

In [ ]:
%%sql
SELECT
 *
FROM
 store

 * postgresql://padwan:***@localhost:5432/pagila
2 rows affected.


store_id  manager_staff_id  address_id               last_update
0         1                 1           1 2022-02-15 09:57:12+00:00
1         2                 2           2 2022-02-15 09:57:12+00:00

In [ ]:
%%sql
SELECT
 *
FROM
 store
INNER JOIN
 customer
ON
 store.store_id=customer.store_id


 * postgresql://padwan:***@localhost:5432/pagila
599 rows affected.


store_id  manager_staff_id  address_id               last_update  \
0           1                 1           1 2022-02-15 09:57:12+00:00   
1           1                 1           1 2022-02-15 09:57:12+00:00   
2           1                 1           1 2022-02-15 09:57:12+00:00   
3           2                 2           2 2022-02-15 09:57:12+00:00   
4           1                 1           1 2022-02-15 09:57:12+00:00   
..        ...               ...         ...                       ...   
594         1                 1           1 2022-02-15 09:57:12+00:00   
595         1                 1           1 2022-02-15 09:57:12+00:00   
596         1                 1           1 2022-02-15 09:57:12+00:00   
597         1                 1           1 2022-02-15 09:57:12+00:00   
598         2                 2           2 2022-02-15 09:57:12+00:00   

     customer_id  store_id first_name  last_name  \
0              1         1       MARY      SMITH   
1              2         1   PATRICIA    JOHNSON   
2              3         1      LINDA   WILLIAMS   
3              4         2    BARBARA      JONES   
4              5         1  ELIZABETH      BROWN   
..           ...       ...        ...        ...   
594          595         1   TERRENCE  GUNDERSON   
595          596         1    ENRIQUE   FORSYTHE   
596          597         1    FREDDIE     DUGGAN   
597          598         1       WADE   DELVALLE   
598          599         2     AUSTIN    CINTRON   

                                     email  address_id  activebool  \
0            MARY.SMITH@sakilacustomer.org           5        True   
1      PATRICIA.JOHNSON@sakilacustomer.org           6        True   
2        LINDA.WILLIAMS@sakilacustomer.org           7        True   
3         BARBARA.JONES@sakilacustomer.org           8        True   
4       ELIZABETH.BROWN@sakilacustomer.org           9        True   
..                                     ...         ...         ...   
594  TERRENCE.GUNDERSON@sakilacustomer.org         601        True   
595    ENRIQUE.FORSYTHE@sakilacustomer.org         602        True   
596      FREDDIE.DUGGAN@sakilacustomer.org         603        True   
597       WADE.DELVALLE@sakilacustomer.org         604        True   
598      AUSTIN.CINTRON@sakilacustomer.org         605        True   

    create_date               last_update  active  
0    2022-02-14 2022-02-15 09:57:20+00:00       1  
1    2022-02-14 2022-02-15 09:57:20+00:00       1  
2    2022-02-14 2022-02-15 09:57:20+00:00       1  
3    2022-02-14 2022-02-15 09:57:20+00:00       1  
4    2022-02-14 2022-02-15 09:57:20+00:00       1  
..          ...                       ...     ...  
594  2022-02-14 2022-02-15 09:57:20+00:00       1  
595  2022-02-14 2022-02-15 09:57:20+00:00       1  
596  2022-02-14 2022-02-15 09:57:20+00:00       1  
597  2022-02-14 2022-02-15 09:57:20+00:00       1  
598  2022-02-14 2022-02-15 09:57:20+00:00       1  

[599 rows x 14 columns]

##1.2 OUTER JOIN

外部結合は、内部結合のようにそれぞれのテーブルの指定したカラム値(列の値)が一致するものを結合するのに加えて、どちらかのテーブルにしか存在しないものに関しても取得します。LEFT  OUTER JOINは左側のテーブルを軸にして外部結合を行い、RIGHT OUTER JOINは右側のテーブルを軸にして外部結合を行います。


###1.2.1 LEFT OUTER JOIN

```
SELECT
 (*)
FROM
 (Table1)
LEFT OUTER JOIN
 (Table2)
ON
 (結合の条件)
```

####例：

In [ ]:
%%sql
SELECT
 fc.category_id, fc.last_update, f.rental_rate, f.rating
FROM
 film_category fc
LEFT OUTER JOIN
 film f
ON
 fc.film_id=f.film_id
ORDER BY
category_id

 * postgresql://padwan:***@localhost:5432/pagila
1000 rows affected.


category_id               last_update rental_rate rating
0              1 2022-02-15 10:07:09+00:00        0.99      R
1              1 2022-02-15 10:07:09+00:00        0.99      G
2              1 2022-02-15 10:07:09+00:00        2.99      R
3              1 2022-02-15 10:07:09+00:00        4.99     PG
4              1 2022-02-15 10:07:09+00:00        2.99  NC-17
..           ...                       ...         ...    ...
995           16 2022-02-15 10:07:09+00:00        0.99  PG-13
996           16 2022-02-15 10:07:09+00:00        2.99  NC-17
997           16 2022-02-15 10:07:09+00:00        2.99     PG
998           16 2022-02-15 10:07:09+00:00        0.99     PG
999           16 2022-02-15 10:07:09+00:00        4.99      G

[1000 rows x 4 columns]

###1.2.2 RIGHT OUTER JOIN

```
SELECT
 (*)
FROM
 (Table1)
RIGHT OUTER JOIN
 (Table2)
ON
 (結合の条件)
```

In [ ]:
%%sql
SELECT
 fc.category_id, fc.last_update, f.rental_rate, f.rating
FROM
 film_category fc
RIGHT OUTER JOIN
 film f
ON
 fc.film_id=f.film_id
ORDER BY
category_id

 * postgresql://padwan:***@localhost:5432/pagila
1000 rows affected.


category_id               last_update rental_rate rating
0              1 2022-02-15 10:07:09+00:00        0.99      R
1              1 2022-02-15 10:07:09+00:00        0.99      G
2              1 2022-02-15 10:07:09+00:00        2.99      R
3              1 2022-02-15 10:07:09+00:00        4.99     PG
4              1 2022-02-15 10:07:09+00:00        2.99  NC-17
..           ...                       ...         ...    ...
995           16 2022-02-15 10:07:09+00:00        0.99  PG-13
996           16 2022-02-15 10:07:09+00:00        2.99  NC-17
997           16 2022-02-15 10:07:09+00:00        2.99     PG
998           16 2022-02-15 10:07:09+00:00        0.99     PG
999           16 2022-02-15 10:07:09+00:00        4.99      G

[1000 rows x 4 columns]

OUTER JOINを使用することで一致しないデータも含めて関連性のあるデータを取得できます。
データの補完や欠損データの確認に役立ちます。

|id|alphabet|    
|---|---|
|1|a|
|2|b|
|3|c|
|4|d|

|id|alphabet|
|---|---|
|1|d|
|2|e|
|5|f|
|6|g|

INNER JOINの場合：

|id|alphabet|id|alphabet|
|---|---|---|---|
|1|a|1|c|
|2|b|2|d|

LEFT OUTER JOINの場合：

|id|alphabet|id|alphabet|
|---|---|---|---|
|1|a|1|d|
|2|b|2|e|
|3|c|
|4|d|

RIGHT OUTER JOINの場合：

|id|alphabet|id|alphabet|
|---|---|---|---|
|1|a|1|d|
|2|b|2|e|
|||5|f|
|||6|g|


##演習問題

問１.
customer テーブルと customer_list テーブルをINNER JOINして、以下の情報を含む結果セットを取得してください。

- 顧客ID (customer_id)
- 顧客の姓(last_name)
- 顧客の名(first_name)
- 顧客の住所 (address)
- 郵便番号 (zip code)
- 電話番号 (phone)
- 都市 (city)
- 国 (country)

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
599 rows affected.


customer_id  last_name first_name                   address  \
0              1      SMITH       MARY            1913 Hanoi Way   
1              2    JOHNSON   PATRICIA          1121 Loja Avenue   
2              3   WILLIAMS      LINDA         692 Joliet Street   
3              4      JONES    BARBARA          1566 Inegl Manor   
4              5      BROWN  ELIZABETH           53 Idfu Parkway   
..           ...        ...        ...                       ...   
594          595  GUNDERSON   TERRENCE       844 Bucuresti Place   
595          596   FORSYTHE    ENRIQUE  1101 Bucuresti Boulevard   
596          597     DUGGAN    FREDDIE    1103 Quilmes Boulevard   
597          598   DELVALLE       WADE       1331 Usak Boulevard   
598          599    CINTRON     AUSTIN      1325 Fukuyama Street   

            phone            city        country  
0     28303384290          Sasebo          Japan  
1    838635286649  San Bernardino  United States  
2    448477190408         Athenai         Greece  
3    705814003527        Myingyan        Myanmar  
4     10655648674          Nantou         Taiwan  
..            ...             ...            ...  
594  935952366111         Jinzhou          China  
595  199514580428          Patras         Greece  
596  644021380889         Sullana           Peru  
597  145308717464        Lausanne    Switzerland  
598  288241215394           Tieli          China  

[599 rows x 7 columns]

問2. customer テーブルと customer_list テーブルとcountryをINNER JOINして、各顧客の次の情報を含む結果セットを取得してください。

- 顧客ID (customer_id)
- 顧客の姓(last_name)
- 顧客の名(first_name)
- 顧客の国名 (country)
- 最終更新日時 (last_update)*斜体テキスト*


In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
599 rows affected.


customer_id  store_id first_name  last_name  \
0              1         1       MARY      SMITH   
1              2         1   PATRICIA    JOHNSON   
2              3         1      LINDA   WILLIAMS   
3              4         2    BARBARA      JONES   
4              5         1  ELIZABETH      BROWN   
..           ...       ...        ...        ...   
594          595         1   TERRENCE  GUNDERSON   
595          596         1    ENRIQUE   FORSYTHE   
596          597         1    FREDDIE     DUGGAN   
597          598         1       WADE   DELVALLE   
598          599         2     AUSTIN    CINTRON   

                                     email  address_id  activebool  \
0            MARY.SMITH@sakilacustomer.org           5        True   
1      PATRICIA.JOHNSON@sakilacustomer.org           6        True   
2        LINDA.WILLIAMS@sakilacustomer.org           7        True   
3         BARBARA.JONES@sakilacustomer.org           8        True   
4       ELIZABETH.BROWN@sakilacustomer.org           9        True   
..                                     ...         ...         ...   
594  TERRENCE.GUNDERSON@sakilacustomer.org         601        True   
595    ENRIQUE.FORSYTHE@sakilacustomer.org         602        True   
596      FREDDIE.DUGGAN@sakilacustomer.org         603        True   
597       WADE.DELVALLE@sakilacustomer.org         604        True   
598      AUSTIN.CINTRON@sakilacustomer.org         605        True   

    create_date               last_update  active  ...  \
0    2022-02-14 2022-02-15 09:57:20+00:00       1  ...   
1    2022-02-14 2022-02-15 09:57:20+00:00       1  ...   
2    2022-02-14 2022-02-15 09:57:20+00:00       1  ...   
3    2022-02-14 2022-02-15 09:57:20+00:00       1  ...   
4    2022-02-14 2022-02-15 09:57:20+00:00       1  ...   
..          ...                       ...     ...  ...   
594  2022-02-14 2022-02-15 09:57:20+00:00       1  ...   
595  2022-02-14 2022-02-15 09:57:20+00:00       1  ...   
596  2022-02-14 2022-02-15 09:57:20+00:00       1  ...   
597  2022-02-14 2022-02-15 09:57:20+00:00       1  ...   
598  2022-02-14 2022-02-15 09:57:20+00:00       1  ...   

                      address zip code         phone            city  \
0              1913 Hanoi Way    35200   28303384290          Sasebo   
1            1121 Loja Avenue    17886  838635286649  San Bernardino   
2           692 Joliet Street    83579  448477190408         Athenai   
3            1566 Inegl Manor    53561  705814003527        Myingyan   
4             53 Idfu Parkway    42399   10655648674          Nantou   
..                        ...      ...           ...             ...   
594       844 Bucuresti Place    36603  935952366111         Jinzhou   
595  1101 Bucuresti Boulevard    97661  199514580428          Patras   
596    1103 Quilmes Boulevard    52137  644021380889         Sullana   
597       1331 Usak Boulevard    61960  145308717464        Lausanne   
598      1325 Fukuyama Street    27107  288241215394           Tieli   

           country   notes sid country_id        country  \
0            Japan  active   1         50          Japan   
1    United States  active   1        103  United States   
2           Greece  active   1         39         Greece   
3          Myanmar  active   2         64        Myanmar   
4           Taiwan  active   1         92         Taiwan   
..             ...     ...  ..        ...            ...   
594          China  active   1         23          China   
595         Greece  active   1         39         Greece   
596           Peru  active   1         74           Peru   
597    Switzerland  active   1         91    Switzerland   
598          China  active   2         23          China   

                  last_update  
0   2022-02-15 09:44:00+00:00  
1   2022-02-15 09:44:00+00:00  
2   2022-02-15 09:44:00+00:00  
3   2022-02-15 09:44:00+00:00  
4   2022-02-15 09:44:00+00:00  
..                        ...  
594 2022-02-15 09:44:00+00:00  


問3. film テーブルと film_actor テーブルをRIGHT OUTER JOINして、以下の情報を含む結果セットを取得してください。

- 映画のタイトル (title)
- 出演した俳優のID (actor_id)
- 最終更新日時 (last_update)

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
5465 rows affected.


title  actor_id               last_update
0          ACADEMY DINOSAUR       1.0 2022-02-15 10:05:03+00:00
1      ANACONDA CONFESSIONS       1.0 2022-02-15 10:05:03+00:00
2               ANGELS LIFE       1.0 2022-02-15 10:05:03+00:00
3     BULWORTH COMMANDMENTS       1.0 2022-02-15 10:05:03+00:00
4             CHEAPER CLYDE       1.0 2022-02-15 10:05:03+00:00
...                     ...       ...                       ...
5460       WARDROBE PHANTOM     200.0 2022-02-15 10:05:03+00:00
5461         WRONG BEHAVIOR     200.0 2022-02-15 10:05:03+00:00
5462       SLACKER LIAISONS       NaN                       NaT
5463            FLIGHT LIES       NaN                       NaT
5464       DRUMLINE CYCLONE       NaN                       NaT

[5465 rows x 3 columns]

問4. film テーブル、film_category テーブル、film_actor テーブル、および language テーブルを結合し、以下の情報を含む結果セットを取得してください。

- 映画のタイトル (film.title)
-カテゴリID (film_category.category_id)
-出演した俳優のID (film_actor.actor_id)
-映画の言語 (language.name)

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
5465 rows affected.


film_title  category_name  actor_id              language  \
0          ACADEMY DINOSAUR              6       1.0  English                
1      ANACONDA CONFESSIONS              2       1.0  English                
2               ANGELS LIFE             13       1.0  English                
3     BULWORTH COMMANDMENTS             10       1.0  English                
4             CHEAPER CLYDE             14       1.0  English                
...                     ...            ...       ...                   ...   
5460       WARDROBE PHANTOM              7     200.0  English                
5461         WRONG BEHAVIOR              3     200.0  English                
5462       SLACKER LIAISONS              7       NaN  English                
5463            FLIGHT LIES             15       NaN  English                
5464       DRUMLINE CYCLONE             16       NaN  English                

                     film_last_update  
0    2022-09-10 16:46:03.905795+00:00  
1    2022-09-10 16:46:03.905795+00:00  
2    2022-09-10 16:46:03.905795+00:00  
3    2022-09-10 16:46:03.905795+00:00  
4    2022-09-10 16:46:03.905795+00:00  
...                               ...  
5460 2022-09-10 16:46:03.905795+00:00  
5461 2022-09-10 16:46:03.905795+00:00  
5462 2022-09-10 16:46:03.905795+00:00  
5463 2022-09-10 16:46:03.905795+00:00  
5464 2022-09-10 16:46:03.905795+00:00  

[5465 rows x 5 columns]

問5.　顧客ごとに、その顧客が支払った合計金額を計算し、顧客IDの昇順で結果を表示して下さい

rentalとpaymentを使います

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
599 rows affected.


customer_id total_payment
0              1       3797.76
1              2       3475.71
2              3       3529.24
3              4       1799.16
4              5       5495.56
..           ...           ...
594          595       3531.00
595          596       2708.16
596          597       2493.75
597          598       1843.16
598          599       1592.39

[599 rows x 2 columns]

#2. VIEW

テーブルの特定部分や複数テーブルを結合し、ユーザに必要な部分だけをあたかも一つのテーブルであるかのように見せるための、仮想的なテーブルの概念です。


CREATE VIEW (ビュー名) AS (定義)

##例:

In [ ]:
%%sql
SELECT
 *
FROM
 customer

 * postgresql://padwan:***@localhost:5432/pagila
599 rows affected.


customer_id  store_id first_name  last_name  \
0              1         1       MARY      SMITH   
1              2         1   PATRICIA    JOHNSON   
2              3         1      LINDA   WILLIAMS   
3              4         2    BARBARA      JONES   
4              5         1  ELIZABETH      BROWN   
..           ...       ...        ...        ...   
594          595         1   TERRENCE  GUNDERSON   
595          596         1    ENRIQUE   FORSYTHE   
596          597         1    FREDDIE     DUGGAN   
597          598         1       WADE   DELVALLE   
598          599         2     AUSTIN    CINTRON   

                                     email  address_id  activebool  \
0            MARY.SMITH@sakilacustomer.org           5        True   
1      PATRICIA.JOHNSON@sakilacustomer.org           6        True   
2        LINDA.WILLIAMS@sakilacustomer.org           7        True   
3         BARBARA.JONES@sakilacustomer.org           8        True   
4       ELIZABETH.BROWN@sakilacustomer.org           9        True   
..                                     ...         ...         ...   
594  TERRENCE.GUNDERSON@sakilacustomer.org         601        True   
595    ENRIQUE.FORSYTHE@sakilacustomer.org         602        True   
596      FREDDIE.DUGGAN@sakilacustomer.org         603        True   
597       WADE.DELVALLE@sakilacustomer.org         604        True   
598      AUSTIN.CINTRON@sakilacustomer.org         605        True   

    create_date               last_update  active  
0    2022-02-14 2022-02-15 09:57:20+00:00       1  
1    2022-02-14 2022-02-15 09:57:20+00:00       1  
2    2022-02-14 2022-02-15 09:57:20+00:00       1  
3    2022-02-14 2022-02-15 09:57:20+00:00       1  
4    2022-02-14 2022-02-15 09:57:20+00:00       1  
..          ...                       ...     ...  
594  2022-02-14 2022-02-15 09:57:20+00:00       1  
595  2022-02-14 2022-02-15 09:57:20+00:00       1  
596  2022-02-14 2022-02-15 09:57:20+00:00       1  
597  2022-02-14 2022-02-15 09:57:20+00:00       1  
598  2022-02-14 2022-02-15 09:57:20+00:00       1  

[599 rows x 10 columns]

In [ ]:
%%sql
CREATE VIEW customer_email AS
SELECT
first_name, last_name, email
FROM
customer

 * postgresql://padwan:***@localhost:5432/pagila
Done.


Empty DataFrame
Columns: []
Index: []

In [ ]:
%%sql
SELECT
*
FROM
customer_name

 * postgresql://padwan:***@localhost:5432/pagila
599 rows affected.


first_name  last_name
0         MARY      SMITH
1     PATRICIA    JOHNSON
2        LINDA   WILLIAMS
3      BARBARA      JONES
4    ELIZABETH      BROWN
..         ...        ...
594   TERRENCE  GUNDERSON
595    ENRIQUE   FORSYTHE
596    FREDDIE     DUGGAN
597       WADE   DELVALLE
598     AUSTIN    CINTRON

[599 rows x 2 columns]

テーブルの内容を変更したい場合、

**CREATE OR REPLACE VIEW ビュー名 AS 定義**

を用いて変更します。

##例:

In [ ]:
%%sql
CREATE OR REPLACE VIEW customer_name AS
SELECT
first_name, last_name, email
FROM
customer;

 * postgresql://padwan:***@localhost:5432/pagila
Done.


Empty DataFrame
Columns: []
Index: []

In [ ]:
%%sql
SELECT
*
FROM
customer_name

 * postgresql://padwan:***@localhost:5432/pagila
599 rows affected.


first_name  last_name                                  email
0         MARY      SMITH          MARY.SMITH@sakilacustomer.org
1     PATRICIA    JOHNSON    PATRICIA.JOHNSON@sakilacustomer.org
2        LINDA   WILLIAMS      LINDA.WILLIAMS@sakilacustomer.org
3      BARBARA      JONES       BARBARA.JONES@sakilacustomer.org
4    ELIZABETH      BROWN     ELIZABETH.BROWN@sakilacustomer.org
..         ...        ...                                    ...
594   TERRENCE  GUNDERSON  TERRENCE.GUNDERSON@sakilacustomer.org
595    ENRIQUE   FORSYTHE    ENRIQUE.FORSYTHE@sakilacustomer.org
596    FREDDIE     DUGGAN      FREDDIE.DUGGAN@sakilacustomer.org
597       WADE   DELVALLE       WADE.DELVALLE@sakilacustomer.org
598     AUSTIN    CINTRON      AUSTIN.CINTRON@sakilacustomer.org

[599 rows x 3 columns]

作ったテーブルを消去するためには、

DROP VIEW テーブル名

で消去します。

##例:


In [ ]:
%%sql
DROP VIEW customer_name

 * postgresql://padwan:***@localhost:5432/pagila
Done.


Empty DataFrame
Columns: []
Index: []

In [ ]:
%%sql
SELECT
*
FROM
customer_name

 * postgresql://padwan:***@localhost:5432/pagila
(psycopg2.errors.UndefinedTable) relation "customer_name" does not exist
LINE 4: customer_name
        ^

[SQL: SELECT
*
FROM
customer_name]
(Background on this error at: https://sqlalche.me/e/20/f405)


##演習問題


以下の列を含む新しいビュー（カスタマービュー）を作成してください。

- 顧客ID (customer_id)
- 顧客の姓(last_name)
- 顧客の名(first_name)
- メールアドレス (email)

それを表示させてください

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
Done.


Empty DataFrame
Columns: []
Index: []

In [ ]:
%%sql
SELECT
*
FROM
customer_view

 * postgresql://padwan:***@localhost:5432/pagila
599 rows affected.


customer_id first_name  last_name                                  email
0              1       MARY      SMITH          MARY.SMITH@sakilacustomer.org
1              2   PATRICIA    JOHNSON    PATRICIA.JOHNSON@sakilacustomer.org
2              3      LINDA   WILLIAMS      LINDA.WILLIAMS@sakilacustomer.org
3              4    BARBARA      JONES       BARBARA.JONES@sakilacustomer.org
4              5  ELIZABETH      BROWN     ELIZABETH.BROWN@sakilacustomer.org
..           ...        ...        ...                                    ...
594          595   TERRENCE  GUNDERSON  TERRENCE.GUNDERSON@sakilacustomer.org
595          596    ENRIQUE   FORSYTHE    ENRIQUE.FORSYTHE@sakilacustomer.org
596          597    FREDDIE     DUGGAN      FREDDIE.DUGGAN@sakilacustomer.org
597          598       WADE   DELVALLE       WADE.DELVALLE@sakilacustomer.org
598          599     AUSTIN    CINTRON      AUSTIN.CINTRON@sakilacustomer.org

[599 rows x 4 columns]

問２　film テーブルから、以下の情報を含む新しいFilmInfoを作成してください。

- 映画のタイトル (title)
- 映画の説明 (description)
- 公開年 (release_year)
- 貸し出し期間 (rental_duration)
- 貸し出し料金 (rental_rate)
- 映画の長さ (length)
- 交換コスト (replacement_cost)
- 映画の評価 (rating)


In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
(psycopg2.errors.DuplicateTable) relation "filminfo" already exists

[SQL: CREATE VIEW FilmInfo AS
SELECT
    title,
    description,
    release_year,
    rental_duration,
    rental_rate,
    length,
    replacement_cost,
    rating
FROM film;]
(Background on this error at: https://sqlalche.me/e/20/f405)


In [ ]:
%%sql
SELECT
*
FROM
FilmInfo

 * postgresql://padwan:***@localhost:5432/pagila
1000 rows affected.


title                                        description  \
0     ACADEMY DINOSAUR  A Epic Drama of a Feminist And a Mad Scientist...   
1       ACE GOLDFINGER  A Astounding Epistle of a Database Administrat...   
2     ADAPTATION HOLES  A Astounding Reflection of a Lumberjack And a ...   
3     AFFAIR PREJUDICE  A Fanciful Documentary of a Frisbee And a Lumb...   
4          AFRICAN EGG  A Fast-Paced Documentary of a Pastry Chef And ...   
..                 ...                                                ...   
995     YOUNG LANGUAGE  A Unbelieveable Yarn of a Boat And a Database ...   
996         YOUTH KICK  A Touching Drama of a Teacher And a Cat who mu...   
997       ZHIVAGO CORE  A Fateful Yarn of a Composer And a Man who mus...   
998  ZOOLANDER FICTION  A Fateful Reflection of a Waitress And a Boat ...   
999          ZORRO ARK  A Intrepid Panorama of a Mad Scientist And a B...   

     release_year  rental_duration rental_rate  length replacement_cost rating  
0            2006                6        0.99      86            20.99     PG  
1            2006                3        4.99      48            12.99      G  
2            2006                7        2.99      50            18.99  NC-17  
3            2006                5        2.99     117            26.99      G  
4            2006                6        2.99     130            22.99      G  
..            ...              ...         ...     ...              ...    ...  
995          2006                6        0.99     183             9.99      G  
996          2006                4        0.99     179            14.99  NC-17  
997          2006                6        0.99     105            10.99  NC-17  
998          2006                5        2.99     101            28.99      R  
999          2006                3        4.99      50            18.99  NC-17  

[1000 rows x 8 columns]

問３. sales_by_store テーブルと staff_list テーブルを結合して、各店舗の売上情報とその店舗の責任者の情報を含む staff_information を作成し、id 列で昇順に並べ替えて下さい。

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
Done.


Empty DataFrame
Columns: []
Index: []

#3. Limit

```
SELECT
(*)
FROM
(テーブル名)
LIMIT
(取得する行数)
```


##例：

In [ ]:
%%sql
SELECT
*
FROM
customer
LIMIT
10

 * postgresql://padwan:***@localhost:5432/pagila
10 rows affected.


customer_id  store_id first_name last_name  \
0            1         1       MARY     SMITH   
1            2         1   PATRICIA   JOHNSON   
2            3         1      LINDA  WILLIAMS   
3            4         2    BARBARA     JONES   
4            5         1  ELIZABETH     BROWN   
5            6         2   JENNIFER     DAVIS   
6            7         1      MARIA    MILLER   
7            8         2      SUSAN    WILSON   
8            9         2   MARGARET     MOORE   
9           10         1    DOROTHY    TAYLOR   

                                 email  address_id  activebool create_date  \
0        MARY.SMITH@sakilacustomer.org           5        True  2022-02-14   
1  PATRICIA.JOHNSON@sakilacustomer.org           6        True  2022-02-14   
2    LINDA.WILLIAMS@sakilacustomer.org           7        True  2022-02-14   
3     BARBARA.JONES@sakilacustomer.org           8        True  2022-02-14   
4   ELIZABETH.BROWN@sakilacustomer.org           9        True  2022-02-14   
5    JENNIFER.DAVIS@sakilacustomer.org          10        True  2022-02-14   
6      MARIA.MILLER@sakilacustomer.org          11        True  2022-02-14   
7      SUSAN.WILSON@sakilacustomer.org          12        True  2022-02-14   
8    MARGARET.MOORE@sakilacustomer.org          13        True  2022-02-14   
9    DOROTHY.TAYLOR@sakilacustomer.org          14        True  2022-02-14   

                last_update  active  
0 2022-02-15 09:57:20+00:00       1  
1 2022-02-15 09:57:20+00:00       1  
2 2022-02-15 09:57:20+00:00       1  
3 2022-02-15 09:57:20+00:00       1  
4 2022-02-15 09:57:20+00:00       1  
5 2022-02-15 09:57:20+00:00       1  
6 2022-02-15 09:57:20+00:00       1  
7 2022-02-15 09:57:20+00:00       1  
8 2022-02-15 09:57:20+00:00       1  
9 2022-02-15 09:57:20+00:00       1

開始位置を変更したい場合は、OFFSETを用いて
```
SELECT
*
FROM
テーブル名
LIMIT
取得する行数
OFFSET
開始位置
```
となります

##例：

In [ ]:
%%sql
SELECT
*
FROM
customer
LIMIT
10
OFFSET
5

 * postgresql://padwan:***@localhost:5432/pagila
10 rows affected.


customer_id  store_id first_name last_name  \
0            6         2   JENNIFER     DAVIS   
1            7         1      MARIA    MILLER   
2            8         2      SUSAN    WILSON   
3            9         2   MARGARET     MOORE   
4           10         1    DOROTHY    TAYLOR   
5           11         2       LISA  ANDERSON   
6           12         1      NANCY    THOMAS   
7           13         2      KAREN   JACKSON   
8           14         2      BETTY     WHITE   
9           15         1      HELEN    HARRIS   

                               email  address_id  activebool create_date  \
0  JENNIFER.DAVIS@sakilacustomer.org          10        True  2022-02-14   
1    MARIA.MILLER@sakilacustomer.org          11        True  2022-02-14   
2    SUSAN.WILSON@sakilacustomer.org          12        True  2022-02-14   
3  MARGARET.MOORE@sakilacustomer.org          13        True  2022-02-14   
4  DOROTHY.TAYLOR@sakilacustomer.org          14        True  2022-02-14   
5   LISA.ANDERSON@sakilacustomer.org          15        True  2022-02-14   
6    NANCY.THOMAS@sakilacustomer.org          16        True  2022-02-14   
7   KAREN.JACKSON@sakilacustomer.org          17        True  2022-02-14   
8     BETTY.WHITE@sakilacustomer.org          18        True  2022-02-14   
9    HELEN.HARRIS@sakilacustomer.org          19        True  2022-02-14   

                last_update  active  
0 2022-02-15 09:57:20+00:00       1  
1 2022-02-15 09:57:20+00:00       1  
2 2022-02-15 09:57:20+00:00       1  
3 2022-02-15 09:57:20+00:00       1  
4 2022-02-15 09:57:20+00:00       1  
5 2022-02-15 09:57:20+00:00       1  
6 2022-02-15 09:57:20+00:00       1  
7 2022-02-15 09:57:20+00:00       1  
8 2022-02-15 09:57:20+00:00       1  
9 2022-02-15 09:57:20+00:00       1

##演習問題

問１. FilmInfoから1から15行目のデータを表示してください。

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
15 rows affected.


title                                        description  \
0   ACADEMY DINOSAUR  A Epic Drama of a Feminist And a Mad Scientist...   
1     ACE GOLDFINGER  A Astounding Epistle of a Database Administrat...   
2   ADAPTATION HOLES  A Astounding Reflection of a Lumberjack And a ...   
3   AFFAIR PREJUDICE  A Fanciful Documentary of a Frisbee And a Lumb...   
4        AFRICAN EGG  A Fast-Paced Documentary of a Pastry Chef And ...   
5       AGENT TRUMAN  A Intrepid Panorama of a Robot And a Boy who m...   
6    AIRPLANE SIERRA  A Touching Saga of a Hunter And a Butler who m...   
7    AIRPORT POLLOCK  A Epic Tale of a Moose And a Girl who must Con...   
8      ALABAMA DEVIL  A Thoughtful Panorama of a Database Administra...   
9   ALADDIN CALENDAR  A Action-Packed Tale of a Man And a Lumberjack...   
10   ALAMO VIDEOTAPE  A Boring Epistle of a Butler And a Cat who mus...   
11    ALASKA PHANTOM  A Fanciful Saga of a Hunter And a Pastry Chef ...   
12        DATE SPEED  A Touching Saga of a Composer And a Moose who ...   
13       ALI FOREVER  A Action-Packed Drama of a Dentist And a Croco...   
14    ALICE FANTASIA  A Emotional Drama of a A Shark And a Database ...   

    release_year  rental_duration rental_rate  length replacement_cost rating  
0           2006                6        0.99      86            20.99     PG  
1           2006                3        4.99      48            12.99      G  
2           2006                7        2.99      50            18.99  NC-17  
3           2006                5        2.99     117            26.99      G  
4           2006                6        2.99     130            22.99      G  
5           2006                3        2.99     169            17.99     PG  
6           2006                6        4.99      62            28.99  PG-13  
7           2006                6        4.99      54            15.99      R  
8           2006                3        2.99     114            21.99  PG-13  
9           2006                6        4.99      63            24.99  NC-17  
10          2006                6        0.99     126            16.99      G  
11          2006                6        0.99     136            22.99     PG  
12          2006                4        0.99     104            19.99      R  
13          2006                4        4.99     150            21.99     PG  
14          2006                6        0.99      94            23.99  NC-17

問２. customerからstore_id,first_name,last_name,emailを抽出して、100行目から150行目まで表示してください

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
50 rows affected.


customer_id  store_id first_name  last_name  \
0           101         1      PEGGY      MYERS   
1           102         1    CRYSTAL       FORD   
2           103         1     GLADYS   HAMILTON   
3           104         1       RITA     GRAHAM   
4           105         1       DAWN   SULLIVAN   
5           106         1     CONNIE    WALLACE   
6           107         1   FLORENCE      WOODS   
7           108         1      TRACY       COLE   
8           109         2       EDNA       WEST   
9           110         2    TIFFANY     JORDAN   
10          111         1     CARMEN      OWENS   
11          112         2       ROSA   REYNOLDS   
12          113         2      CINDY     FISHER   
13          114         2      GRACE      ELLIS   
14          115         1      WENDY   HARRISON   
15          116         1   VICTORIA     GIBSON   
16          117         1      EDITH   MCDONALD   
17          118         1        KIM       CRUZ   
18          119         1     SHERRY   MARSHALL   
19          120         2     SYLVIA      ORTIZ   
20          121         1  JOSEPHINE      GOMEZ   
21          122         1     THELMA     MURRAY   
22          123         2    SHANNON    FREEMAN   
23          124         1     SHEILA      WELLS   
24          125         1      ETHEL       WEBB   
25          126         1      ELLEN    SIMPSON   
26          127         2     ELAINE    STEVENS   
27          128         1   MARJORIE     TUCKER   
28          129         1     CARRIE     PORTER   
29          130         1  CHARLOTTE     HUNTER   
30          131         2     MONICA      HICKS   
31          132         2     ESTHER   CRAWFORD   
32          133         1    PAULINE      HENRY   
33          134         1       EMMA       BOYD   
34          135         2    JUANITA      MASON   
35          136         2      ANITA    MORALES   
36          137         2     RHONDA    KENNEDY   
37          138         1      HAZEL     WARREN   
38          139         1      AMBER      DIXON   
39          140         1        EVA      RAMOS   
40          141         1     DEBBIE      REYES   
41          142         1      APRIL      BURNS   
42          143         1     LESLIE     GORDON   
43          144         1      CLARA       SHAW   
44          145         1    LUCILLE     HOLMES   
45          146         1      JAMIE       RICE   
46          147         2     JOANNE  ROBERTSON   
47          148         1    ELEANOR       HUNT   
48          149         1    VALERIE      BLACK   
49          150         2   DANIELLE    DANIELS   

                                  email  address_id  activebool create_date  \
0        PEGGY.MYERS@sakilacustomer.org         105        True  2022-02-14   
1       CRYSTAL.FORD@sakilacustomer.org         106        True  2022-02-14   
2    GLADYS.HAMILTON@sakilacustomer.org         107        True  2022-02-14   
3        RITA.GRAHAM@sakilacustomer.org         108        True  2022-02-14   
4      DAWN.SULLIVAN@sakilacustomer.org         109        True  2022-02-14   
5     CONNIE.WALLACE@sakilacustomer.org         110        True  2022-02-14   
6     FLORENCE.WOODS@sakilacustomer.org         111        True  2022-02-14   
7         TRACY.COLE@sakilacustomer.org         112        True  2022-02-14   
8          EDNA.WEST@sakilacustomer.org         113        True  2022-02-14   
9     TIFFANY.JORDAN@sakilacustomer.org         114        True  2022-02-14   
10      CARMEN.OWENS@sakilacustomer.org         115        True  2022-02-14   
11     ROSA.REYNOLDS@sakilacustomer.org         116        True  2022-02-14   
12      CINDY.FISHER@sakilacustomer.org         117        True  2022-02-14   
13       GRACE.ELLIS@sakilacustomer.org         118        True  2022-02-14   
14    WENDY.HARRISON@sakilacustomer.org         119        True  2022-02-14   
15   VICTORIA.GIBSON@sakilacustomer.org         120        True  2022-02-14   
16    EDITH.MCDONALD@sakilacustomer.org         121        

#総合問題

以下の列を含む新しいテーブル（film_view）を作成してください。

- 映画ID (film_id)
- タイトル (title)
- 説明 (description)
- カテゴリ名（film_category テーブルと結合してカテゴリ名を取得）
- 言語名（language テーブルと結合して言語名を取得）
- レンタル料金 (rental_rate)
- 再生時間 (length)
- 交換費用 (replacement_cost)
- 評価 (rating)
- 最終更新日時 (last_update)

In [ ]:
%%sql

 * postgresql://padwan:***@localhost:5432/pagila
Done.


Empty DataFrame
Columns: []
Index: []

In [ ]:
%%sql
SELECT
*
FROM
film_view

 * postgresql://padwan:***@localhost:5432/pagila
(psycopg2.errors.UndefinedTable) relation "film_view" does not exist
LINE 4: film_view
        ^

[SQL: SELECT
*
FROM
film_view]
(Background on this error at: https://sqlalche.me/e/20/f405)


問2film_view ビューから、以下の条件を持つ映画を抽出してください。

- レンタル料金 (rental_rate) が2.99以上の映画。
- 映画の長さ (length) が120分以上の映画。
- 類似性の高い映画カテゴリ (category_id) が複数存在する映画。
- 総合評価 (rating) が 'PG-13' または 'R' の映画。
- 抽出する映画の数を5本に制限。

In [ ]:
%%sql


UsageError: Cell magic `%%sql` not found.
